pkg/ddl/index.go

```diff
@@ -598,6 +675,264 @@ func decodeAddIndexArgs(job *model.Job) (
+			InitAndAddColumnToTable(tblInfo, hiddenCol)
+		}
+	}
+	if err = checkAddColumnTooManyColumns(len(tblInfo.Columns)); err != nil {
+		job.State = model.JobStateCancelled
+		return nil, errors.Trace(err)
+	}
+	indexInfo, err = BuildIndexInfo(
+		nil,
+		tblInfo,
+		indexName,
+		isPK,
+		unique,
+		isVector,
+		indexPartSpecifications,
+		indexOption,
+		model.StateNone,
+	)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return nil, errors.Trace(err)
+	}
+	if isPK {
+		if _, err = CheckPKOnGeneratedColumn(tblInfo, indexPartSpecifications); err != nil {
+			job.State = model.JobStateCancelled
+			return nil, err
+		}
+	}
+	indexInfo.ID = AllocateIndexID(tblInfo)
+	tblInfo.Indices = append(tblInfo.Indices, indexInfo)
+	if err = checkTooManyIndexes(tblInfo.Indices); err != nil {
+		job.State = model.JobStateCancelled
+		return nil, errors.Trace(err)
+	}
+	// Here we need do this check before set state to `DeleteOnly`,
+	// because if hidden columns has been set to `DeleteOnly`,
+	// the `DeleteOnly` columns are missing when we do this check.
+	if err := checkInvisibleIndexOnPK(tblInfo); err != nil {
+		job.State = model.JobStateCancelled
+		return nil, err
+	}
+	logutil.DDLLogger().Info("[ddl] run add index job", zap.String("job", job.String()), zap.Reflect("indexInfo", indexInfo))
+	return indexInfo, nil
+}
+
+func (w *worker) onCreateVectorIndex(jobCtx *jobContext, t *meta.Meta, job *model.Job) (ver int64, err error) {
+	// Handle the rolling back job.
+	if job.IsRollingback() {
+		ver, err = onDropIndex(jobCtx, t, job)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+		return ver, nil
+	}
+
+	// Handle normal job.
+	schemaID := job.SchemaID
+	tblInfo, err := GetTableInfoAndCancelFaultJob(t, job, schemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	if err := checkTableTypeForVectorIndex(tblInfo); err != nil {
+		return ver, errors.Trace(err)
+	}
+	if tblInfo.TiFlashReplica == nil || tblInfo.TiFlashReplica.Count == 0 {
+		return ver, dbterror.ErrUnsupportedAddVectorIndex.FastGenByArgs("unsupported empty TiFlash replica, the replica is nil")
+	}
+
+	var (
+		indexName              model.CIStr
+		indexOption            *ast.IndexOption
+		indexPartSpecification *ast.IndexPartSpecification
+		funcExpr               string
+	)
+	err = job.DecodeArgs(&indexName, &indexPartSpecification, &indexOption, &funcExpr)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+	indexPartSpecification.Expr, err = generatedexpr.ParseExpression(funcExpr)
+	if err != nil {
+		job.State = model.JobStateCancelled
+		return ver, errors.Trace(err)
+	}
+	defer func() {
+		indexPartSpecification.Expr = nil
+	}()
+
+	indexInfo, err := checkAndBuildIndexInfo(job, tblInfo, indexName, false, false, true, []*ast.IndexPartSpecification{indexPartSpecification}, indexOption, nil)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+	originalState := indexInfo.State
+	switch indexInfo.State {
```

Comment:
+1

Reply:
Although this can be considered to remove DO and WO, I think WR still needs to be added to inform TiFlash to fill the index (this action may take a long time if the middle owner changes, etc.). 
However, the two states of DO and WO are not time-consuming, and may be supported later, and some index processing can be handled uniformly. Therefore, I will keep this for now.

Reply:
> DO and WO are not time-consuming

But there is no benefit?

> may be supported later

Is there any plan that support writing a vector index? We should add these intermediate states at that time.

> some index processing can be handled uniformly

What do you mean “uniformly", I can only see there is a lot of boilerplate code.

Reply:
> > DO and WO are not time-consuming
> 
> But there is no benefit?

Yes.



> > some index processing can be handled uniformly
> 
> What do you mean “uniformly", I can only see there is a lot of boilerplate code.

For example, there are DO and WO states when dropping vector index, but not when adding. Is there something strange